pageserver/src/layered_repository.rs

```diff
@@ -1162,22 +1162,34 @@ impl LayeredTimeline {
         // check, though. We should also aim at flushing layers that consume
         // a lot of memory and/or aren't receiving much updates anymore.
         let mut disk_consistent_lsn = last_record_lsn;
-        while let Some(oldest_layer) = layers.peek_oldest_open() {
-            // Does this layer need freezing?
+
+        while let Some((oldest_layer, oldest_generation)) = layers.peek_oldest_open() {
             let oldest_pending_lsn = oldest_layer.get_oldest_pending_lsn();
-            let distance = last_valid_lsn.0 - oldest_pending_lsn.0;
-            if !force && distance < OLDEST_INMEM_DISTANCE {
+
+            // Does this layer need freezing?
+            //
+            // Write out all in-memory layers that contain WAL older than OLDEST_INMEM_DISTANCE.
+            // Or if 'force' is true, write out all of them. If we reach a layer with the same
+            // generation number, we know that we have cycled through all layers that were open
+            // when we started. We don't want to process layers inserted after we started, to
+            // avoid getting into an infinite loop trying to process again entries that we
+            // inserted ourselves.
+            let distance:i128 = i128::from(last_record_lsn.0) - i128::from(oldest_pending_lsn.0);
+            if distance < 0
+                || (!force && distance < i128::from(OLDEST_INMEM_DISTANCE))
```

Comment:
Do we need i128?

Reply:
I used i128 to avoid integer overflow/underflow. `oldest_pending_lsn` can be larger than `last_record_lsn` and I wanted distance to become negative in that case. I'm not totally happy with this solution. There are a few different ways this could be written, using checked_sub() or saturating_sub(), but I'm not sure which one I like the most. Perhaps this would be better?

```
            let distance = last_record_lsn.0.saturating_sub(oldest_pending_lsn.0);
            if oldest_pending_lsn.0 > last_record_lsn.0
                || (!force && distance < OLDEST_INMEM_DISTANCE)
                || oldest_generation == current_generation
            {
                info!(
                    "the oldest layer is now {} which is {} bytes behind last_record_lsn",
                    oldest_layer.filename().display(),
                    distance
                );
                disk_consistent_lsn = oldest_pending_lsn;
                break;
            }
```
Or this (note the difference in the info-message, too) :
```
            if distance.is_none()   // oldest_pending_lsn > last_record_lsn
                || (!force && distance.unwrap_or(0) < OLDEST_INMEM_DISTANCE)
                || oldest_generation == current_generation
            {
                info!(
                    "the oldest layer is now {} which is {} bytes behind last_record_lsn",
                    oldest_layer.filename().display(),
                    distance.unwrap_or(0)
                );
                disk_consistent_lsn = oldest_pending_lsn;
                break;
            }
```

Reply:
Or perhaps we should add a function into the Lsn type for this?

Reply:
I see, i128 approach seems fair for the most accurate log right?

Reply:
> I see, i128 approach seems fair for the most accurate log right?

Yeah.

I added a `widening_sub` function to Lsn type, and used that here. I like that best so far. It's the same i128 approach, but the helper function makes it look less cluttered here.